lib/provisioning/ProvisionDispatcher.js

```diff
@@ -242,103 +281,111 @@ class ProvisionDispatcher {
+                this._onOwnerChange();
+            }, (err, children) => {
+                if (err) {
+                    this._log.error('list failed',
+                                    { zkPath: `${this._zkEndpoint}${zkPath}`,
+                                      error: err });
+                    return;
+                }
+                this._owners = children;
+                this._log.debug('owners updated',
+                                { zkPath: `${this._zkEndpoint}${zkPath}`,
+                                  owners: this._owners });
+                if (this._isLeader) {
+                    this._redispatchProvisions();
+                }
+            })
+        );
     }
 
     _onMyselfChange(cb) {
         // monitor change in my content
         const myPath = `${OWNERS}/${this._myName}`;
-        this._client.getData(
-            myPath, event => {
-                this._log.debug('got owner self event',
-                    { zkPath: this._zkEndpoint + myPath,
-                        event });
-                this._onMyselfChange(cb);
-            },
-            (err, data) => {
-                if (err &&
-                    err.getCode() !== zookeeper.Exception.NO_NODE) {
-                    this._log.error('error in getData',
-                        { zkPath: this._zkEndpoint + myPath,
-                            error: err });
-                    return cb(err);
-                }
-                if (data !== undefined) {
-                    const provisionList = JSON.parse(data);
-                    this._log.info('provisioning update',
-                        { zkPath: this._zkEndpoint + myPath,
-                            provisionList });
-                    return cb(null, provisionList);
+        this._withRandomDelayIfTest(
+            () => this._client.getData(
+                myPath, event => {
+                    this._log.debug('got owner self event',
+                                    { zkPath: this._zkEndpoint + myPath,
+                                      event });
+                    this._onMyselfChange(cb);
+                },
+                (err, data) => {
+                    if (err &&
+                        err.getCode() !== zookeeper.Exception.NO_NODE) {
+                        this._log.error('error in getData',
+                                        { zkPath: this._zkEndpoint + myPath,
+                                          error: err });
+                        return cb(err);
+                    }
+                    if (data !== undefined) {
+                        const provisionList = JSON.parse(data);
+                        this._log.info('provisioning update',
+                                       { zkPath: this._zkEndpoint + myPath,
+                                         provisionList });
+                        return cb(null, provisionList);
+                    }
+                    return undefined;
```

Comment:
Is there a case where `data` is defined and we will not be returning the callback?

Reply:
Ah yeah undefined. If we expect to hit that case, we should return callback for that case, an empty one should be fine.

Reply:
No we don't want to call the callback because there's no change in provisioning (again this is not a regular callback, rather an event notification system).